# Rossmann Sales Data Cleaning
**Author:** Priyanshu Singh | Data Analyst  
**Dataset:** Rossmann Store Sales (Kaggle)  
**Tools:** Python, Pandas, PostgreSQL
## What this covers
This project cleans and preps the Rossmann retail sales dataset, just over a million records across 1,115 stores in Germany, so it's ready for analysis and forecasting.
## Steps
1. Load and explore the raw data
2. Fix data types and extract date features
3. Remove invalid records using business logic
4. Clean store metadata and handle missing values
5. Merge datasets and validate
6. Load clean data to PostgreSQL

In [1]:
import pandas as pd

## Raw Data
Two raw files: train.csv (the sales records) and store.csv (store metadata).
train.csv has 1,017,209 daily sales records across 1,115 stores. store.csv covers each store's type, assortment, and competition details.

In [2]:
df_train = pd.read_csv('../data/raw/train.csv', low_memory = False)
df_store = pd.read_csv('../data/raw/store.csv')

In [3]:
print(df_train.shape)

(1017209, 9)


In [4]:
df_train.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday
0,1,5,2015-07-31,5263,555,1,1,0,1
1,2,5,2015-07-31,6064,625,1,1,0,1
2,3,5,2015-07-31,8314,821,1,1,0,1
3,4,5,2015-07-31,13995,1498,1,1,0,1
4,5,5,2015-07-31,4822,559,1,1,0,1


In [5]:
print(df_train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB
None


## Fix Data Types
Date is stored as an object (text) instead of a datetime, so the first step is converting it. That gives us access to Year, Month, Day, and WeekOfYear, which the forecasting models need.

In [6]:
df_train['Date'] = pd.to_datetime(df_train['Date'])

In [7]:
df_train["Date"].dtype

dtype('<M8[ns]')

In [8]:
df_train['Date'].sample(10)

498014    2014-04-11
311733    2014-10-12
798505    2013-07-16
819872    2013-06-26
508489    2014-04-02
1006683   2013-01-10
805079    2013-07-10
663739    2013-11-14
348683    2014-09-02
463844    2014-05-12
Name: Date, dtype: datetime64[ns]

In [9]:
df_train['Year'] = df_train['Date'].dt.year
df_train['Month'] = df_train['Date'].dt.month
df_train['WeekOfYear'] = df_train['Date'].dt.isocalendar().week
df_train['Day'] = df_train['Date'].dt.day

In [10]:
df_train.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday,Year,Month,WeekOfYear,Day
0,1,5,2015-07-31,5263,555,1,1,0,1,2015,7,31,31
1,2,5,2015-07-31,6064,625,1,1,0,1,2015,7,31,31
2,3,5,2015-07-31,8314,821,1,1,0,1,2015,7,31,31
3,4,5,2015-07-31,13995,1498,1,1,0,1,2015,7,31,31
4,5,5,2015-07-31,4822,559,1,1,0,1,2015,7,31,31


In [11]:
print(df_train.isnull().sum())

Store            0
DayOfWeek        0
Date             0
Sales            0
Customers        0
Open             0
Promo            0
StateHoliday     0
SchoolHoliday    0
Year             0
Month            0
WeekOfYear       0
Day              0
dtype: int64


In [12]:
print(df_train.describe())

              Store     DayOfWeek                           Date  \
count  1.017209e+06  1.017209e+06                        1017209   
mean   5.584297e+02  3.998341e+00  2014-04-11 01:30:42.846061824   
min    1.000000e+00  1.000000e+00            2013-01-01 00:00:00   
25%    2.800000e+02  2.000000e+00            2013-08-17 00:00:00   
50%    5.580000e+02  4.000000e+00            2014-04-02 00:00:00   
75%    8.380000e+02  6.000000e+00            2014-12-12 00:00:00   
max    1.115000e+03  7.000000e+00            2015-07-31 00:00:00   
std    3.219087e+02  1.997391e+00                            NaN   

              Sales     Customers          Open         Promo  SchoolHoliday  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   1.017209e+06   
mean   5.773819e+03  6.331459e+02  8.301067e-01  3.815145e-01   1.786467e-01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
25%    3.727000e+03  4.050000e+02  1.000000e+00  0.000000e+00   0.0

In [13]:
df_train[df_train['Open'] == 0].shape

(172817, 13)

In [14]:
df_train[df_train['Sales'] ==0].shape

(172871, 13)

## Remove Invalid Records
**Business Logic Applied:**
- Dropped 172,817 rows where Open = 0. Closed stores don't have sales, so they're dead weight for forecasting. Also dropped 54 rows where Open = 1 but Sales = 0, since an open store with zero sales is almost certainly bad data, not a real business day.
Both calls are just applying how retail actually works, not some universal rule.

In [15]:
df_train = df_train[df_train['Open'] != 0] 
print(df_train.shape)

(844392, 13)


In [16]:
df_train = df_train[~((df_train['Open'] ==1) & (df_train['Sales'] ==0))]
print(df_train.shape)

(844338, 13)


In [17]:
# IQR sanity check on sales after removing invalid records
Q1 = df_train['Sales'].quantile(0.25)
Q3 = df_train['Sales'].quantile(0.75)
IQR = Q3 - Q1
outliers = df_train[
    (df_train['Sales'] < Q1 - 1.5 * IQR) |
    (df_train['Sales'] > Q3 + 1.5 * IQR)
]
print(f"Sales outlier rows: {len(outliers)} of {len(df_train)} ({len(outliers)/len(df_train)*100:.1f}%)")
print(outliers[['Store', 'Date', 'Sales', 'Customers']].head(10))

Sales outlier rows: 30769 of 844338 (3.6%)
     Store       Date  Sales  Customers
3        4 2015-07-31  13995       1498
6        7 2015-07-31  15344       1414
23      24 2015-07-31  14190       1082
24      25 2015-07-31  14180       1586
83      84 2015-07-31  14949       1439
107    108 2015-07-31  14927        992
124    125 2015-07-31  18227       2041
191    192 2015-07-31  16191       1027
206    207 2015-07-31  13957       1436
210    211 2015-07-31  17286       1659


Dropped 172,817 rows where Open = 0. Closed stores don't sell anything, so they're just noise for a forecasting model.

The other drop was smaller: 54 rows where Open = 1 but Sales = 0. A store that's open with zero sales usually isn't a real business day, it's bad data.

Neither call is some universal rule. It's just how retail works.

In [18]:
print(df_train.duplicated().sum())

0


In [19]:
print(df_train['Open'].unique())
print(df_train['Promo'].unique())

[1]
[1 0]


In [20]:
df_train.columns = df_train.columns.str.lower()

In [21]:
df_train.columns

Index(['store', 'dayofweek', 'date', 'sales', 'customers', 'open', 'promo',
       'stateholiday', 'schoolholiday', 'year', 'month', 'weekofyear', 'day'],
      dtype='object')

In [22]:
print(df_train.isnull().sum())
print(df_train.shape)
print(df_train.dtypes)

store            0
dayofweek        0
date             0
sales            0
customers        0
open             0
promo            0
stateholiday     0
schoolholiday    0
year             0
month            0
weekofyear       0
day              0
dtype: int64
(844338, 13)
store                     int64
dayofweek                 int64
date             datetime64[ns]
sales                     int64
customers                 int64
open                      int64
promo                     int64
stateholiday             object
schoolholiday             int64
year                      int32
month                     int32
weekofyear               UInt32
day                       int32
dtype: object


In [23]:
df_train.to_csv('../data/clean/train_clean.csv', index=False)

In [24]:
df_store.head()

,Store,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,c,a,1270.0,9.0,2008.0,0,NaN,NaN,NaN
1,2,a,a,570.0,11.0,2007.0,1,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,a,a,14130.0,12.0,2006.0,1,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,c,c,620.0,9.0,2009.0,0,NaN,NaN,NaN
4,5,a,a,29910.0,4.0,2015.0,0,NaN,NaN,NaN


In [25]:
df_store.describe()

,Store,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear
count,1115.00000,1112.000000,761.000000,761.000000,1115.000000,571.000000,571.000000
mean,558.00000,5404.901079,7.224704,2008.668857,0.512108,23.595447,2011.763573
std,322.01708,7663.174720,3.212348,6.195983,0.500078,14.141984,1.674935
min,1.00000,20.000000,1.000000,1900.000000,0.000000,1.000000,2009.000000
25%,279.50000,717.500000,4.000000,2006.000000,0.000000,13.000000,2011.000000
50%,558.00000,2325.000000,8.000000,2010.000000,1.000000,22.000000,2012.000000
75%,836.50000,6882.500000,10.000000,2013.000000,1.000000,37.000000,2013.000000
max,1115.00000,75860.000000,12.000000,2015.000000,1.000000,50.000000,2015.000000


In [26]:
df_store.isnull().sum()

Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            3
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64

In [27]:
df_store.columns = df_store.columns.str.lower()

## Clean Store Metadata
store.csv has gaps in the competition and promotion columns.
- CompetitionDistance gets the median. One store reports 75,860m of competition distance against a median of 2,325m, that kind of outlier would drag a -mean fill way off.
- CompetitionOpenSinceMonth and CompetitionOpenSinceYear are filled with the median too, since both need to land on whole numbers, a month from 1 to 12, a believable year.
- Promo2SinceWeek, Promo2SinceYear, and PromoInterval just get 0. These nulls aren't a data problem, they're stores where Promo2 = 0 and no promotion ever ran, so there's nothing to fill in.

In [42]:
# check whether missing CompetitionDistance values pile up in one store type,
# because if they do, a plain median fill could skew that group
missing_comp = df_store[df_store['competitiondistance'].isna()]
print("Missing CompetitionDistance by StoreType:")
print(missing_comp['storetype'].value_counts())
print(f"\nTotal missing: {missing_comp.shape[0]} of {df_store.shape[0]} stores")

# flag which rows were imputed before filling, so anyone using this data later
# can tell real values apart from filled ones
df_store['competitiondistance_imputed'] = df_store['competitiondistance'].isna().astype(int)

Missing CompetitionDistance by StoreType:
Series([], Name: count, dtype: int64)

Total missing: 0 of 1115 stores


Only 3 of 1,115 stores are missing CompetitionDistance. Two are Type d, 
one is Type a.  
That's too small a sample to say it clusters by store type.  
Median fill is fine for a gap this small. The imputed flag marks the three rows so downstream analysis can drop or check them later if needed.

In [29]:
median_column = ['competitiondistance', 'competitionopensincemonth', 'competitionopensinceyear']
for col in median_column:
    df_store[col] = df_store[col].fillna(df_store[col].median())

In [30]:
df_store.columns

Index(['store', 'storetype', 'assortment', 'competitiondistance',
       'competitionopensincemonth', 'competitionopensinceyear', 'promo2',
       'promo2sinceweek', 'promo2sinceyear', 'promointerval',
       'competitiondistance_imputed'],
      dtype='object')

In [31]:
zero_columns = ['promo2sinceweek', 'promo2sinceyear', 'promointerval']
for col in zero_columns:
    df_store[col] = df_store[col].fillna(0)

In [32]:
print(df_store.isnull().sum())

store                          0
storetype                      0
assortment                     0
competitiondistance            0
competitionopensincemonth      0
competitionopensinceyear       0
promo2                         0
promo2sinceweek                0
promo2sinceyear                0
promointerval                  0
competitiondistance_imputed    0
dtype: int64


In [33]:
df_store.duplicated().sum()

np.int64(0)

In [34]:
print(df_store.isnull().sum())
print(df_store.shape)
print(df_store.dtypes)

store                          0
storetype                      0
assortment                     0
competitiondistance            0
competitionopensincemonth      0
competitionopensinceyear       0
promo2                         0
promo2sinceweek                0
promo2sinceyear                0
promointerval                  0
competitiondistance_imputed    0
dtype: int64
(1115, 11)
store                            int64
storetype                       object
assortment                      object
competitiondistance            float64
competitionopensincemonth      float64
competitionopensinceyear       float64
promo2                           int64
promo2sinceweek                float64
promo2sinceyear                float64
promointerval                   object
competitiondistance_imputed      int64
dtype: object


In [35]:
type_conv = ['competitionopensincemonth', 'competitionopensinceyear', 'promo2sinceweek', 'promo2sinceyear']
for col in type_conv:
    df_store[col] = df_store[col].astype(int)

print(df_store.dtypes)

store                            int64
storetype                       object
assortment                      object
competitiondistance            float64
competitionopensincemonth        int64
competitionopensinceyear         int64
promo2                           int64
promo2sinceweek                  int64
promo2sinceyear                  int64
promointerval                   object
competitiondistance_imputed      int64
dtype: object


In [36]:
df_store.to_csv('../data/clean/store_clean.csv', index=False)

## Step 5: Merge Datasets
Combining train.csv and store.csv on the Store column with a left join.  
This brings the store metadata, type, competition info, promotions, onto every sales record, so we can dig into the business questions properly.

In [37]:
df = pd.merge(df_train,df_store,on ='store',how='left')

In [38]:
print(df.shape)
df.head()

(844338, 23)


,store,dayofweek,date,sales,customers,open,promo,stateholiday,schoolholiday,year,...,storetype,assortment,competitiondistance,competitionopensincemonth,competitionopensinceyear,promo2,promo2sinceweek,promo2sinceyear,promointerval,competitiondistance_imputed
0,1,5,2015-07-31,5263,555,1,1,0,1,2015,...,c,a,1270.0,9,2008,0,0,0,0,0
1,2,5,2015-07-31,6064,625,1,1,0,1,2015,...,a,a,570.0,11,2007,1,13,2010,"Jan,Apr,Jul,Oct",0
2,3,5,2015-07-31,8314,821,1,1,0,1,2015,...,a,a,14130.0,12,2006,1,14,2011,"Jan,Apr,Jul,Oct",0
3,4,5,2015-07-31,13995,1498,1,1,0,1,2015,...,c,c,620.0,9,2009,0,0,0,0,0
4,5,5,2015-07-31,4822,559,1,1,0,1,2015,...,a,a,29910.0,4,2015,0,0,0,0,0


In [39]:
df.isnull().sum()

store                          0
dayofweek                      0
date                           0
sales                          0
customers                      0
open                           0
promo                          0
stateholiday                   0
schoolholiday                  0
year                           0
month                          0
weekofyear                     0
day                            0
storetype                      0
assortment                     0
competitiondistance            0
competitionopensincemonth      0
competitionopensinceyear       0
promo2                         0
promo2sinceweek                0
promo2sinceyear                0
promointerval                  0
competitiondistance_imputed    0
dtype: int64

In [40]:
df.to_csv('../data/clean/merged_cleaned.csv',index =False)

## Load to PostgreSQL
Loading clean merged data to PostgreSQL for SQL analysis.

In [41]:
from dotenv import load_dotenv
import os
from sqlalchemy import create_engine
import pandas as pd

load_dotenv()
password = os.getenv('DB_PASSWORD')
engine = create_engine(f'postgresql://postgres:{password}@localhost:5432/rossmann')

df = pd.read_csv('../data/clean/merged_cleaned.csv', low_memory=False)
df_store = pd.read_csv('../data/clean/store_clean.csv')

df.to_sql('rossmann_sales', engine, if_exists='replace', index=False)
df_store.to_sql('rossmann_store', engine, if_exists='replace', index=False)

print("Data loaded successfully")

Data loaded successfully
